### ISSUE TITLE
a question about latency

### ISSUE THREAD

**Issue by xnxky** [shyiko/mysql-binlog-connector-java#75]
Hi, thanks again for this great tool. Here is my scenario:

We have aws rds mysql master (which is row based replication) and an ec2 instance running the mysql-binlog-connector (bandwidth and throughput are not an issue) reading the binlog from the mysql master. We just extracted the timestamp out of the binlog event, and calculate the difference between the current time and timestamp. The number of events is around 1k per second.

However, the difference, for p95, is about 1.3 seconds, and the max one could be as much as 8~9 seconds. We are hoping to have it under 100ms. Do you have an estimate or benchmark about how much latency this tool has in the best case?

Thanks.

**Comment by shyiko** [comment]
Hi. 

The best way to find where the bottleneck is (**in your particular case**) is obviously to take a look in profiler, but the usual suspect is AbstractRowsEventDataDeserializer. For example, if you are using <=0.2.4 and your data stream contains **a lot** of `DATETIME|TIMESTAMP|...` switching to unreleased 1.0.0 `*RowsEventDataDeserializer`s might result in ~300% performance improvement (this is actually the reason why we are switching to `long`s from `Date|Timestamp`) (see [changelog.md](https://github.com/shyiko/mysql-binlog-connector-java/blob/master/changelog.md) for more). 

The good news you can provide your own `EventDataDeserializer`s (without the need to change anything in `mysql-binlog-connector-java`) to optimize deserialization based on what you need in the end (for example you might not need anything but "id"s in case of *RowsEventData and so you could set custom `EventDataDeserializer` which would skip all but row ids). The bed news - it really depends on what your replication stream "is made of".

tl;dr I expect most of the time to be spent either reading from the network or deserializing the data stream (see [this package](https://github.com/shyiko/mysql-binlog-connector-java/tree/master/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization)). So I would look into those two first. I would also check IO on MySQL server just to be sure that delay is not caused by stuff like sync_binlog=1, low innodb_max_dirty_pages_pct and so on.

**Comment by xnxky** [comment]
Hi, Stanley

      Thanks a lot for your reply. I modified your code to add some timing metrics in datadog. In our case, the P99 network IO (in.read ) is about 50ms and the P99 time between 2 IOs is about 90 ~ 100ms. (P99 of the difference between the current time and the timestamp is about 4 seconds). The P99 difference between the start time of network IO read and corresponding binlog timestamps is close to 4 seconds (so it accounts for most of the latency.) I'm not an expert in msyql internals(not in charge of them either), so what do you think could be the root cause of it? Really appreciate your help.

**Comment by xnxky** [comment]
Hi, Stanley

       Thanks a lot for your reply. I modified your code to add some timing metrics in datadog. The P99 network IO (in.read) is about 50ms and the P99 time between 2 IOs is about 90 ~ 100ms. (The P99 difference between the current time and the timestamp is about 4 seconds.) Further investigatiosn indicates that the P99 difference between the start time of a network IO and corresponding binlog timestamps is close to 4 seconds (so it account for most of the latency.). And we are quite sure that we are processing much faster than the pace of mysql generating binlogs, so when it's stable in each read we get all the binlogs available there. I'm not an expert in mysql internals and what do you think are the causes of it? Really appreciate your help.

**Comment by shyiko** [comment]
Hi.

If I understand you correctly time between [this](https://github.com/xnxky/mysql-binlog-connector-java/commit/aed0c061b2f60818f8e01e7c6aed90be96e2c9c4#diff-7addc4e3eed6e9254e088abc015ac8adR590) and [this](https://github.com/xnxky/mysql-binlog-connector-java/commit/aed0c061b2f60818f8e01e7c6aed90be96e2c9c4#diff-7addc4e3eed6e9254e088abc015ac8adR604) is ~4s (correct me if I'm wrong). If that's the case - try connecting remotely using [VisualVM](https://visualvm.java.net/) (or similar) to see what is taking so long (4s is insane, especially for a single event).

**Comment by xnxky** [comment]
Hi, Stanley

         Sorry I did not make it clear. That part you pointed out is only about 50ms. The 4 
seconds is the difference is between [this](https://github.com/xnxky/mysql-binlog-connector-java/blob/master/src/main/java/com/github/shyiko/mysql/binlog/io/BufferedSocketInputStream.java#L55) and the timestamp of the binlog events read by [this](https://github.com/xnxky/mysql-binlog-connector-java/blob/master/src/main/java/com/github/shyiko/mysql/binlog/io/BufferedSocketInputStream.java#L57). Thanks again for your help.
